### ISSUE TITLE
pipe/thread function

### ISSUE THREAD

**Issue by sethyuan** [sethyuan/func.js#1]
Because there is no macro or operator overloading support in JavaScript, function has to be considered a possible implementation for piping/threading.

E.g.

``` js
// arr will be [, ]
const arr = pipe(
  [1, 2, 3, 4],
  _ => map(x => 2*x, _),
  _ => filter(x => x % 4 === 0, _),
  _ => new Array(_.length)
);
```

**Comment by sethyuan** [comment]
Another way to do this is through convention, like this:

``` js
let _;
_ = [1, 2, 3, 4];
_ = map(x => 2*x, _);
_ = filter(x => x % 4 === 0, _);
_ = new Array(_.length);
const arr = _;
```

This is supposed to be more performant, because no functions need to be created, but it is less readable.

**Comment by sethyuan** [comment]
Implemented in 1531172
